textile/features.textile

```diff
@@ -1305,6 +1307,23 @@ h4. Message
 ** @(TM2e)@ @encoding@ string
 ** @(TM2i)@ @extras@ JSON-encodable object, used to contain any arbitrary key value pairs which may also contain other primitive JSON types, JSON-encodable objects or JSON-encodable arrays. The @extras@ field is provided to contain message metadata and/or ancillary payloads in support of specific functionality, e.g. push. Each of these supported extensions is documented separately; for 1.1 the only supported extension is @push@, via the @extras.push@ member; 1.2 adds the @delta@ extension whose keys and values are described by the attributes of the type @DeltaExtras@, and the @headers@ extension, which contains arbitrary @string->string@ key-value pairs, settable at publish time, and @ref@ whose keys and values are described by the attributes of the type @ReferenceExtras@. Unless otherwise specified, the client library should not attempt to do any filtering or validation of the @extras@ field itself, but should treat it opaquely, encoding it and passing it to realtime unaltered.
 ** @(TM2f)@ @timestamp@ time in milliseconds since epoch. If a message received from Ably does not contain a @timestamp@, it should be set to the @timestamp@ of the encapsulating @ProtocolMessage@
+** @(TM2j)@ @action@ enum - the type of the message. Any message received from Ably must have an @action@ field set to one of the following values:
+*** @(TM2j1)@ 0: @MESSAGE_UNSET@
+*** @(TM2j2)@ 1: @MESSAGE_CREATE@
+*** @(TM2j3)@ 2: @MESSAGE_UPDATE@
+*** @(TM2j4)@ 3: @MESSAGE_DELETE@
+*** @(TM2j5)@ 4: @ANNOTATION_CREATE@
+*** @(TM2j6)@ 5: @ANNOTATION_DELETE@
+*** @(TM2j7)@ 6: @META_OCCUPANCY@
+** @(TM2K)@ @serial@ string - the origin timeserial of the message. This will be set, after publish, to the @channelSerial@ of the @ProtocolMessage@ that encapsulated the @Message@, with an additional offset that locates the @Message@ in the @ProtocolMessage@ messages array. If the message is received from Ably, this must be set.
```

Comment:
Remember that this is the SDK features spec, it specifies how SDK's should be implemented.

Put yourself in the position of an SDK developer reading this (and the rest of the TM2 series). Your job is to make your SDK compliant with these requirements. How do I do that? "This will be set, after publish, to the channelSerial of the ProtocolMessage..." -- ok, how do I do that? After publish as in after it's been sent on the wire? But by then I can't modify it. Is it telling me to set it? If it's received from ably it "must" be set? So what do I do if it isn't?

Answer: you can't, and actually there is no requirement for the SDK implementation here other than to expose the `serial`. Plus a brief explanation for a docstring or whatever. `* @(TM2k)@ @serial@ string, an opaque string that uniquely identifies this message` would do fine

Reply:
Thanks for the feedback Simon, this is really useful! I'll update the message string too